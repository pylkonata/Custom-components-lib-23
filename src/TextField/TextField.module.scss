.wrapper {
  width: 13.75rem;
  display: inline-flex;
  flex-direction: column;
  margin: 0.5rem;
}

.container {
  height: 3.5rem;
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  position: relative;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375rem;
  letter-spacing: 0.00938rem;
  cursor: text;
  padding: 0;
}

.input {
  width: 100%;
  height: 100%;
  display: block;
  box-sizing: border-box;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-size: 1rem;
  line-height: 1.44rem;
  letter-spacing: inherit;
  color: currentColor;
  background: none;
  border: 0;
  margin: 0;
  outline: 0;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.label {
  max-width: 85%;
  top: 50%;
  position: absolute;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.44rem;
  letter-spacing: 0.00938rem;
  color: rgba(0, 0, 0, 0.6);
  padding: 0 2px;
  margin: 0;
  transform: translateY(-50%);
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: none;
  z-index: 999;

  &.focused {
    font-size: 0.8125rem;
    z-index: 1000;
  }

  &:has(+ .input:focus-visible) {
    color: #1976d2;
  }

  &.errorLabel:has(+ .input:focus-visible),
  &.errorLabel {
    color: #d32f2f;
  }
}

.errorText {
  color: #d32f2f;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.2;
  letter-spacing: 0.03333rem;
  text-align: left;
  margin: 0.1875rem 0.875rem 0;
  word-wrap: break-word;
}

.outlined {
  border: 1px solid rgba(0, 0, 0, 0.23);
  color: rgba(0, 0, 0, 0.87);
  border-radius: 0.25rem;

  &:hover {
    border: 1px solid rgba(0, 0, 0, 0.87);
    color: #000000;
  }

  &:has(.input:focus-visible) {
    border: 2px solid #1976d2;
  }

  &.error:has(.input:focus-visible) {
    border: 2px solid #d32f2f;
  }

  &.error,
  &:hover .error {
    border: 1px solid #d32f2f;
  }

  .cover {
    position: absolute;
    left: 0.5rem;
    top: -0.1rem;
    height: 0.125rem;
    margin: 0;
    padding: 0;
    opacity: 1;
    visibility: hidden;
  }

  .coverFocused {
    visibility: visible;
    background-color: white;
    z-index: 999;
  }

  .input {
    padding: 1rem 0.875rem;
  }

  .label {
    left: 1rem;
    transition: top 0.2s, left 0.2s;

    &.focused {
      left: 0.75rem;
      top: -50%;
      transform: translateY(65%);
    }
  }
}

.filled {
  background-color: rgba(0, 0, 0, 0.06);
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-bottom: 1px solid rgba(0, 0, 0, 0.42);

  &:hover {
    background-color: rgba(0, 0, 0, 0.09);
    border-bottom: 1px solid rgba(0, 0, 0, 0.87);
  }

  .input {
    padding: 1.5625rem 0.75rem 0.5rem;
  }

  .label {
    left: 0.75rem;
    transition: top 0.2s, left 0.2s;

    &.focused {
      top: 0.9rem;
    }
  }

  &:has(.input:focus-visible) {
    border-bottom: 2px solid #1976d2;
    transition: border-bottom-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  }

  &.error:has(.input:focus-visible) {
    border-bottom: 2px solid #d32f2f;
  }

  &.error,
  &:hover .error {
    border-bottom: 1px solid #d32f2f;
  }
}

.standard {
  height: 3rem;
  border-bottom: 1px solid rgba(0, 0, 0, 0.42);

  &:hover {
    border-bottom: 2px solid rgba(0, 0, 0, 0.87);
  }

  .input {
    padding: 4px 0 5px;
    margin-top: 1rem;
  }

  .label {
    top: 2rem;
    left: 0;
    transition: top 0.2s, left 0.2s;

    &.focused {
      top: 0.5rem;
    }
  }

  &:has(.input:focus-visible) {
    border-bottom: 2px solid #1976d2;
    transition: border-bottom-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  }

  &.error:has(.input:focus-visible) {
    border-bottom: 2px solid #d32f2f;
  }

  &.error,
  &:hover .error {
    border-bottom: 1px solid #d32f2f;
  }
}

.standard+.errorText {
  margin: 0.1875rem 0 0;
}

.disabled {
  pointer-events: none;

  &.filled,
  &.standard {
    border-bottom-style: dotted;
  }

  &.filled {
    background-color: rgba(0, 0, 0, 0.12);
  }

  &.label {
    color: rgba(0, 0, 0, 0.38);
  }
}
