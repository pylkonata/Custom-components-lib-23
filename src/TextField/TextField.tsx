import { ChangeEvent, useEffect, useId, useRef, useState } from 'react';
import s from './TextField.module.scss';

type VariantTextField = 'outlined' | 'filled' | 'standard';
interface ITextFieldProps {
	label: string;
	variant: VariantTextField;
	error?: boolean;
	errorText?: string;
	disabled?: boolean;
}

const TextField = ({
	label,
	variant,
	error,
	errorText,
	disabled = false,
	...props
}: ITextFieldProps) => {
	const id = useId();
	const [focusState, setFocusState] = useState(false);
	const [value, setValue] = useState('');
	const [labelWidth, setLabelWidth] = useState(0);
	const labelRef = useRef<HTMLLabelElement>(null);

	useEffect(() => {
		setLabelWidth(labelRef.current ? labelRef.current.getBoundingClientRect().width : 0);
	}, [labelRef]);

	const onChangeInput = (e: ChangeEvent<HTMLInputElement>) => {
		setValue(e.target.value);
	};
	const containerClasses = [
		s['container'],
		s[variant],
		error ? s['error'] : '',
		disabled ? s['disabled'] : '',
	].join(' ');

	const labelClasses = [
		s['label'],
		focusState ? s['focused'] : '',
		error ? s['errorLabel'] : '',
	].join(' ');

	return (
		<div className={s['wrapper']} data-testid='textField'>
			<div className={containerClasses} data-testid='container'>
				<label className={labelClasses} htmlFor={id} ref={labelRef}>
					{error ? 'Error' : label}
				</label>
				<input
					data-testid='input'
					id={id}
					className={s.input}
					type='text'
					disabled={disabled}
					value={value}
					onFocus={() => {
						setFocusState(true);
					}}
					onBlur={() => {
						setFocusState(Boolean(value));
					}}
					onChange={onChangeInput}
					{...props}
				/>
				{variant === 'outlined' && (
					<p
						className={`${focusState ? `${s['coverFocused']} ` : ''} ${s['cover']}`}
						style={{ width: labelWidth }}
						data-testid='cover'
					/>
				)}
			</div>
			{errorText && <p className={s['errorText']}>{errorText}</p>}
		</div>
	);
};

export default TextField;
