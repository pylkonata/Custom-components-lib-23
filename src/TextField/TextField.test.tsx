import { render, screen, fireEvent } from '@testing-library/react';
import TextField from './TextField';

describe('TextField', () => {
	test('render basic TextField', () => {
		render(<TextField label='Label text' variant='outlined' />);
		const textField = screen.getByTestId('textField');
		const cover = screen.queryByTestId('cover');
		expect(cover).toBeInTheDocument();
		expect(textField).toBeInTheDocument();
		expect(textField).toMatchSnapshot();
		expect(textField).toBeEnabled();
	});

	test('render TextField without cover element', () => {
		render(<TextField label='Label text' variant='filled' />);
		const cover = screen.queryByTestId('cover');
		expect(cover).not.toBeInTheDocument();
	});

	test('check input value', () => {
		render(<TextField label='Name' variant='standard' />);
		function hasInputValue(e: Element, inputValue: string) {
			return screen.getByDisplayValue(inputValue) === e;
		}
		const input = screen.getByTestId('input');
		expect(input).toBeInTheDocument();
		expect(input).toContainHTML('');

		fireEvent.input(input, { target: { value: 'test 123' } });
		expect(screen.getByTestId('input')).toContainHTML('test 123');

		fireEvent.change(input, { target: { value: 'new value' } });
		expect(hasInputValue(input, 'new value')).toBe(true);
	});

	test('render TextField with disabled value', () => {
		render(<TextField label='disabled' variant='outlined' disabled={true} />);

		const input = screen.getByTestId('input');
		const container = screen.getByTestId('container');

		expect(input).toBeInTheDocument();
		expect(input).toBeDisabled();

		expect(container).toBeInTheDocument();
		expect(container).toHaveClass('container', 'outlined', 'disabled');
	});

	test('render TextField with error value', () => {
		render(<TextField label='error' variant='filled' error={true} errorText='Error text' />);

		const errorText = screen.getByText('Error text');
		const container = screen.getByTestId('container');

		expect(errorText).toBeInTheDocument();

		expect(container).toBeInTheDocument();
		expect(container).toHaveClass('container', 'filled', 'error');
	});
});
