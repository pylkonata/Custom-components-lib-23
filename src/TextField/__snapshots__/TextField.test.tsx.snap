// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField render basic TextField 1`] = `
<div
  class="wrapper"
  data-testid="textField"
>
  <div
    class="container outlined  "
    data-testid="container"
  >
    <label
      class="label  "
      for=":r0:"
    >
      Label text
    </label>
    <input
      class="input"
      data-testid="input"
      id=":r0:"
      type="text"
      value=""
    />
    <p
      class=" cover"
      data-testid="cover"
      style="width: 0px;"
    />
  </div>
</div>
`;
